### ISSUE TITLE
Question: how to specify KeyMod of Control for all platforms?

### ISSUE THREAD

**Issue by psobolewskiPhD** [pyapp-kit/app-model#229]
KeyMod.CtrlCmd results in Ctrl on windows (and linux?) but Cmd on macOS
KeyMod.WinCtrl says it does Meta on Windows and Ctrl on macOS

Is there a way to specify a Control-Foo binding that uses Control on all platforms?
KeyMod.Ctrl does't exist, but there is KeyCode.Ctrl

Based on 
https://github.com/pyapp-kit/app-model/blob/a0d95b19c0e8d61ccab6bea69d1adace910d9157/src/app_model/backends/qt/_qkeymap.py#L220-L238

it looks like it wouldn't be easy to have two different KeyMod map to the same QCTRL...

**Comment by tlambert03** [comment]
> Is there a way to specify

can you clarify where exactly you're specifying these?  You might be looking a bit low in the codebase...

My expectation was that, most of the time, keybindings would be declared as keybinding rules (either during the declaration of a plugin or in an Action for an application).  is that what you're trying to do here?  or are you doing something special at a lower level

**Comment by tlambert03** [comment]
a general answer to your question is that [KeybindingRules](https://app-model.readthedocs.io/en/latest/reference/types/#app_model.types.KeyBindingRule) allow for platform specific logic.   So, if just saying `KeyBindingRule(primary='ctrl+X')` doesn't do what you need, you can do `KeyBindingRule(primary='ctrl+X', mac='ctrl+X', linux='ctrl+X',  win='ctrl+X')`.

since this all follows the vscode specification for declaring keybindings, you might find it useful to see how a vscode app/extensions declares stuff.  Here's an example with a ton of keybindings declared: 

https://github.com/microsoft/vscode-sublime-keybindings/blob/5d7eac6603cf57b8a8f1b5cff6f63c1f59529111/package.json#L262

I can also write up a short example of how you could test stuff out locally a little more directly if that would be helpful

**Comment by psobolewskiPhD** [comment]
Well, specifically here:
https://github.com/napari/napari/blob/b2edccd6e40e04467ccfeec0257c2160783f7187/napari/utils/shortcuts.py#L12
```
    'napari:delete_selected_layers': [KeyMod.CtrlCmd | KeyCode.Delete],
```
That resolves to Control-Del everywhere, but Command-Del on macOS which is a system keybind. The idea was to make it Control-Del everywhere, but `KeyMod.Ctrl` doesn't work. I tried `KeyCode.Ctrl` but that made the binding resolve to PageDown on macOS for some reason!

**Comment by tlambert03** [comment]
I see, that list has lost the ability to declare platform specificity.  Well, how about you try WinCtrl (regardless of what my comment said), and if it doesn't do what you need, you might need to extend the concept of your default key bindings to allow for platform specificity

**Comment by psobolewskiPhD** [comment]
I think it's best to make that a 2nd key binding. Thanks for the insight -- I will close as an explicit KeyMod.Ctrl isn't going to be  a thing.

**Comment by tlambert03** [comment]
> an explicit KeyMod.Ctrl isn't going to be a thing.

I'm not 100% sure about this yet :)  reopening

**Comment by psobolewskiPhD** [comment]
Keep! Sorry!
❤️

**Comment by tlambert03** [comment]
one additional thing I noted:  if you use the string form: `ctrl+X` ... then you will get ctrl on both platforms (I think ... please double check).  It's just the integer form using a KeyMod enum that doesn't have an "always ctrl" variant

**Comment by psobolewskiPhD** [comment]
Maybe i'm dense, but:
`'napari:delete_selected_layers': [ctrl+Delete],`
doesn't work:
`NameError: name 'ctrl' is not defined`

**Comment by tlambert03** [comment]
indeed.  it would need to be a string :)  `'ctrl+Delete'`

**Comment by tlambert03** [comment]
let's back up a bit... keys can be encoded as strings `'ctrl+alt+delete` or as integers using KeyMod and KeyCode: `KeyMod.CtrlCmd | KeyMod.Alt | KeyCode.Delete` ... 

you definitely _can't_ combine the two, so you can't do `'ctrl' + Key.Delete` and you can't do `KeyMod.CtrlCmd + 'delete'`, or you'll get the operand error you showed above.

when it comes time to actually convert those to an OS-bound keybinding, it will pass through `Keybinding.validate`:

https://github.com/pyapp-kit/app-model/blob/a0d95b19c0e8d61ccab6bea69d1adace910d9157/src/app_model/types/_keys/_keybindings.py#L295-L305

and there, it will receive slightly different behavior depending on if it's in a string or integer form:

https://github.com/pyapp-kit/app-model/blob/a0d95b19c0e8d61ccab6bea69d1adace910d9157/src/app_model/types/_keys/_keybindings.py#L71-L93

note that the `CtrlCmd` logic is only differentiated in the int form (since it's an int)... but if you want always `ctrl`, i think you can just use `'ctrl'`

so,  without looking deeper at how napari actually processes those lists you have there, I would expect you should be able to use:

```
'napari:delete_selected_layers': ['ctrl+delete'],
```

**Comment by tlambert03** [comment]
> but Command-Del on macOS which is a system keybind and as a result don't work. The idea was to make it Control-Del everywhere,

I would actually dig into this a bit more... i would 100% expect to be able to delete a layer with cmd+del on macos... and `ctrl+del` feels like an odd workaround.  why can't you use cmd+del?  what error exactly do you hit?

**Comment by psobolewskiPhD** [comment]
There is no error, I assume it's processed by the OS.

**Comment by tlambert03** [comment]
i guess I mean, what do you mean when you say "as a result don't work"  ... what is the behavior?  does nothing at all?

**Comment by psobolewskiPhD** [comment]
Yeah, nothing. If I bind Shift-Del or whatever it works. But Command-Del doesn't -- I assume because it's a system binding.

**Comment by tlambert03** [comment]
yeah: what I'm saying is that ... if it's just an assumption, before you commit to an unusual non-standard keybinding... seems like it would be worth answering that question convincingly.  it would actually surprise me if, in Qt, you can't bind cmd-del to anything you want

**Comment by psobolewskiPhD** [comment]
I almost wonder if it's related to backspace vs delete?
macOS keyboard key is labeled `delete` but behaves like backspace....

**Comment by tlambert03** [comment]
it would be super helpful to reduce the problem to a mwe... rather than going through the entire napari app.  if you can do that and demonstrate no Qt response, i'd be happy to look into it

**Comment by psobolewskiPhD** [comment]
Boom. If i change it to :     
`'napari:delete_selected_layers': [KeyMod.CtrlCmd | KeyCode.Backspace],`
it works

**Comment by tlambert03** [comment]
https://github.com/pyapp-kit/app-model/blob/a0d95b19c0e8d61ccab6bea69d1adace910d9157/src/app_model/types/_keys/_key_codes.py#L104

https://github.com/pyapp-kit/app-model/blob/a0d95b19c0e8d61ccab6bea69d1adace910d9157/src/app_model/types/_keys/_key_codes.py#L92

:)

**Comment by psobolewskiPhD** [comment]
Yup! if I go back to the original and use `Fn-Command-delete-key` it works.
So everything is working correctly in principle.
I do think it would be useful to be able to explicitly bind Control-foo on macOS. macOS does have Control-bar bindings -- mostly for text handling that I know, but still.

**Comment by tlambert03** [comment]
> I do think it would be useful to be able to explicitly bind Control-foo on macOS.

you can: with strings... 'ctrl+backspace'.  
as I look into this, one reason I think they probably avoided having *both* the more useful `KeyMod.CtrlCmd` (which does the think you usually want it to) *and* `KeyMod.Ctrl` (which also seems rather straightforward, but would require lots of conditionals elsewhere if it were the only one available) is that if both are present, it gets a little less unambiguous when these are used as bit flags in a chain: `KeyMod.CtrlCmd | KeyMod.Ctrl | KeyMod.X` ... it introduces overlapping masks

**Comment by psobolewskiPhD** [comment]
Makes sense.
In my defense -- 🤣 -- part of my confusion comes from:
`'napari:delete_selected_points': [
        KeyCode.Digit1,
        KeyCode.Delete,
        KeyCode.Backspace,
    ],`
So here the button I push on my keyboard (labeled delete, but working like backspace) works, but in the GUI we show just the first 2, so I didn't realize why it worked.

**Comment by tlambert03** [comment]
I'm not quite sure exactly what napari does with that list: `[ KeyCode.Digit1, KeyCode.Delete, KeyCode.Backspace, ]` ... is that a sequence of key presses?  alternate key presses?

**Comment by psobolewskiPhD** [comment]
It's 3 bindings. but the gui only shows the first two:
<img width="670" alt="image" src="https://github.com/user-attachments/assets/03e7036d-3990-43af-9d81-a5448c714316" />
And because attention to detail is not my thing, I didn't notice the icon is actually forward delete...

**Comment by psobolewskiPhD** [comment]
> it gets a little less unambiguous when these are used as bit flags in a chain: KeyMod.CtrlCmd | KeyMod.Ctrl | KeyMod.X ... it introduces overlapping masks

So how would you encode that without `KeyMod.Ctrl`?
On macOS `Command-Control-Foo` is a legit combo for bindings.

**Comment by tlambert03** [comment]
> I guess this would be KeyMod.CtrlCmd | KeyMod.WinCtrl | Foo?

yep
